pubsub-mapped-api/src/test/java/com/google/pubsub/clients/consumer/KafkaConsumerTest.java

```diff
@@ -0,0 +1,333 @@
+    assertTrue(topics.equals(subscribed));
+  }
+
+  @Test
+  public void testUnsubscribe() {
+    grpcServerRule.getServiceRegistry().addService(new SubscriberImpl());
+
+    consumer.subscribe(Arrays.asList("topic2", "topic1"));
+    consumer.unsubscribe();
+
+    assertTrue(consumer.subscription().isEmpty());
+  }
+
+  @Test
+  public void testResubscribe() {
+    grpcServerRule.getServiceRegistry().addService(new SubscriberImpl());
+
+    Set<String> topics = new HashSet<>(Arrays.asList("topic4", "topic3"));
+
+    consumer.subscribe(Arrays.asList("topic2", "topic1"));
+    consumer.subscribe(Arrays.asList("topic3", "topic4"));
+
+    Set<String> subscribed = consumer.subscription();
+    assertTrue(topics.equals(subscribed));
+  }
+
+  @Test
+  public void testSubscribeWithRecurringTopics() {
+    grpcServerRule.getServiceRegistry().addService(new SubscriberImpl());
+
+    Set<String> topics = new HashSet<>(Arrays.asList("topic", "topic1"));
+
+    consumer.subscribe(Arrays.asList("topic", "topic1", "topic"));
+
+    Set<String> subscribed = consumer.subscription();
+    assertTrue(topics.equals(subscribed));
+  }
+
+  @Test
+  public void testUnsubscribeSubscriptionDeleted() {
+    grpcServerRule.getServiceRegistry().addService(new SubscriberImpl());
+
+    Set<String> topics = new HashSet<>(Arrays.asList("topic", "topic1"));
+
+    consumer.subscribe(Arrays.asList("topic", "topic1", "topic"));
+
+    Set<String> subscribed = consumer.subscription();
+    assertTrue(topics.equals(subscribed));
+  }
+
+  @Test
+  public void testPatternSubscription() {
+    grpcServerRule.getServiceRegistry().addService(new SubscriberImpl());
+    grpcServerRule.getServiceRegistry().addService(new PublisherImpl());
+
+    Pattern pattern = Pattern.compile("[a-z]*\\d{3}cat");
+
+    consumer.subscribe(pattern, null);
+    Set<String> subscribed = consumer.subscription();
+    Set<String> expectedTopics = new HashSet<>(Arrays.asList("thisis123cat",
+        "funnycats000cat"));
+    assertEquals(expectedTopics, subscribed);
+  }
+
+  @Test
+  public void emptyPatternFails() {
+    try {
+      Pattern pattern = null;
+      consumer.subscribe(pattern, null);
+      fail();
+    } catch (IllegalArgumentException e) {
+      assertEquals("Topic pattern to subscribe to cannot be null", e.getMessage());
+    }
+  }
+
+  @Test
+  public void nullTopicListFails() {
+    try {
+      List<String> topics = null;
+      consumer.subscribe(topics);
+      fail();
+    } catch (IllegalArgumentException e) {
+      assertEquals("Topic collection to subscribe to cannot be null", e.getMessage());
+    }
+  }
+
+  @Test
+  public void emptyTopicFails() {
+    try {
+      List<String> topics = new ArrayList<>(Collections.singletonList("    "));
+      consumer.subscribe(topics);
+      fail();
+    } catch (IllegalArgumentException e) {
+      assertEquals("Topic collection to subscribe to cannot contain null or empty topic",
```

Comment:
same comment about line wrapping.  Here, keeping all the arguments on the same line is clearer